src/Framework/TaskParameterEventArgs.cs

```diff
@@ -0,0 +1,212 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using Microsoft.Build.Shared;
+
+namespace Microsoft.Build.Framework
+{
+    public enum TaskParameterMessageKind
+    {
+        TaskInput,
+        TaskOutput,
+        AddItem,
+        RemoveItem,
+    }
+
+    /// <summary>
+    /// This class is used by tasks to log their parameters (input, output).
+    /// The intrinsic ItemGroupIntrinsicTask to add or remove items also
+    /// uses this class.
+    /// </summary>
+    public class TaskParameterEventArgs : BuildMessageEventArgs
+    {
+        /// <summary>
+        /// Creates an instance of this class for the given task parameter.
+        /// </summary>
+        public TaskParameterEventArgs
+        (
+            TaskParameterMessageKind kind,
+            string itemName,
+            IList items,
+            bool logItemMetadata,
+            DateTime eventTimestamp
+        )
+            : base(null, null, null, MessageImportance.Low, eventTimestamp)
+        {
+            Kind = kind;
+            ItemName = itemName;
+            Items = items;
+            LogItemMetadata = logItemMetadata;
+        }
+
+        public TaskParameterMessageKind Kind { get; private set; }
+        public string ItemName { get; private set; }
+        public IList Items { get; private set; }
+        public bool LogItemMetadata { get; private set; }
+
+        /// <summary>
+        /// The <see cref="TaskParameterEventArgs"/> type is declared in Microsoft.Build.Framework
+        /// which is a declarations assembly. The logic to realize the Message is in Microsoft.Build
+        /// which is an implementations assembly. This seems like the easiest way to inject the
+        /// implementation for realizing the Message.
+        /// </summary>
+        internal static Func<TaskParameterEventArgs, string> MessageGetter = args => null;
```

Comment:
Is there a way in which this will cause null pointer exceptions or the breaking change where API users got the message before whereas now they'll get null? Some scenarios I'm thinking about:
- Is there a way to log this event without having the static fields initialized? Like a project without any actual tasks, just target itemgroups.
- If other API users replay the binlog, they won't be able to rebuild the message since there's no TaskExecutionHost to init the statics.

Reply:
This is a very good point and I'll need to think about ways to ensure these are initialized reliably. I had to touch the type manually from a unit-test and I didn't like it.

Reply:
So, the logging part is easy. To log a TaskParameterEventArgs you have to go through ItemGroupLoggingHelper. Item group add/remove also goes through it.

Reply:
I came up with nothing better than calling the static ctor from BinaryLogReplayEventArgs :) It's a little smelly but does the job.